02_knight_and_princess/robert_gaisin/main.cpp

```diff
@@ -0,0 +1,164 @@
+#include <ctime>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+using VertexId = int;
+using EdgeId = int;
+
+std::ostream& operator<<(std::ostream& out, const vector<int>& int_vector) {
+  for (auto it = int_vector.begin(); it != int_vector.end(); ++it) {
+    out << *it;
+    if (it != int_vector.end() - 1) {
+      out << ", ";
+    }
+  }
+  return out;
+}
+
+struct Vertex {
+  const VertexId id = 0;
+  vector<EdgeId> edge_ids;
+};
+
+std::ostream& operator<<(std::ostream& out, const Vertex& vertex) {
+  out << "{" << endl;
+  out << "      \"id\": " << vertex.id << "," << endl;
+  out << "      \"edge_ids\": [" << vertex.edge_ids << "]" << endl;
+  out << "}";
+  return out;
+}
+
+struct Edge {
+  const EdgeId id = 0;
+  const VertexId begin = 0;
+  const VertexId end = 0;
+};
+
+std::ostream& operator<<(std::ostream& out, const Edge& edge) {
+  out << "{" << endl;
+  out << "      \"id\": " << edge.id << "," << endl;
+  out << "      \"vertex_ids\": [" << edge.begin << ", " << edge.end << "]"
+      << endl;
+  out << "}";
+  return out;
+}
+
+class Graph {
+ public:
+  void add_vertex();
+
+  void add_edge(const VertexId& begin, const VertexId& end);
+
+  const vector<Vertex>& vertices() const { return vertices_; }
+  const vector<Edge>& edges() const { return edges_; }
+
+ private:
+  VertexId num_of_vrt_ = 0;
+  EdgeId num_of_edg_ = 0;
+
+  VertexId next_vertex_id() { return num_of_vrt_++; }
+  EdgeId next_edge_id() { return num_of_edg_++; }
+
+  bool connection(const VertexId& begin, const VertexId& end);
+  vector<Vertex> vertices_;
+  vector<Edge> edges_;
+};
+
+bool Graph::connection(const VertexId& begin, const VertexId& end) {
+  bool ret = false;
+  for (const EdgeId& edge_num : vertices_[begin].edge_ids) {
+    if ((ret = edges_[edge_num].begin == end || edges_[edge_num].end == end)) {
+      break;
+    }
+  }
+  return ret;
+}
+
+void Graph::add_vertex() {
+  Vertex vertex = {next_vertex_id()};
+  vertices_.push_back(vertex);
+}
+
+void Graph::add_edge(const VertexId& begin, const VertexId& end) {
+  if ((begin < num_of_vrt_) && (end < num_of_vrt_) && !connection(begin, end)) {
+    Edge edge = {next_edge_id(), begin, end};
+    vertices_[begin].edge_ids.push_back(edge.id);
+    vertices_[end].edge_ids.push_back(edge.id);
+    edges_.push_back(edge);
+  }
```

Comment:
Во-первых: Всегда лучше использовать ранний выход из функции:
```suggestion
  if (begin >= num_of_vrt_ || end >= num_of_vrt_ || connection(begin, end)) {
    return;
  }
  Edge edge = {next_edge_id(), begin, end};
  vertices_[begin].edge_ids.push_back(edge.id);
  vertices_[end].edge_ids.push_back(edge.id);
  edges_.push_back(edge);
```

Во-вторых: так как у нас нет инпута от юзера или из сети, то нет смысла обрабатывать ошибки. В данном случае лучше использовать `assert`, они работают только во время дебага, а в production коде отключаются, и как раз нужны для проверки самого себя на адекватность:
```cpp
assert(!connection(begin, end) && "Vertices already connected");
```

В-третьих: По какой такой логике, ты решил что `id` можно проверять на валидность по количеству элементов? А если эта логика изменится? Лучше всегда писать максимально стабильные проверки, которые не несут в себе никаких предположений, соответственно, нужно пройтись по массиву вершин, и проверить что там существуют вершины с такими `id`